### PR TITLE
End2end replication 2-phase commit integration

### PR DESCRIPTION
**Pull request by parkerduckworth**
### What's being changed:

Implementation of the replication 2-phase commit protocol between the coordinator and participants

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: usecases/sharding/replicated_index.go [hunk 1]
```diff
@@ -13,24 +13,36 @@ package sharding
...
+		docIDs []uint64, dryRun bool) replica.SimpleResponse
+	ReplicateReferences(ctx context.Context, shard, requestID string,
+		refs []objects.BatchReference) replica.SimpleResponse
+	CommitReplication(ctx context.Context, shard,
+		requestID string) interface{}
+	AbortReplication(ctx context.Context, shardName,
+		requestID string) interface{}
 }
 
 type ReplicatedIndex struct {
```
Comment (redouan-rhazouani) [inline usecases/sharding/replicated_index.go:48]: ReplicatedIndex is not directly related to replication hence we can safely move it to another location.
I think a better place would be `repos/db` where the Replicator can be implement by `DB type` in replication.go

**Review by redouan-rhazouani — CHANGES_REQUESTED** [review]

File: adapters/clients/replication.go [hunk 2]
```diff
@@ -155,7 +158,7 @@ func (c *ReplicationClient) MergeObject(ctx context.Context, host, index, shard,
 }
 
 func (c *ReplicationClient) AddReferences(ctx context.Context, host, index,
-	shard, requestID string, refs []objects.BatchReference,
+	shard, requestID string, refs objects.BatchReferences,
```
[the remark this answers is not in the document]
Comment (parkerduckworth) [reply adapters/clients/replication.go:161]: Sure, will make the change

File: adapters/repos/db/index.go [hunk 3]
```diff
@@ -1285,20 +1308,50 @@ func (i *Index) IncomingDeleteObjectBatch(ctx context.Context, shardName string,
...
+	return (*Index)(ri).ReplicateUpdate(ctx, shardName, requestID, merge)
+}
+
+func (ri *replicatedIndex) ReplicateDeletion(ctx context.Context, shardName,
+	requestID string, uuid strfmt.UUID,
+) entrep.SimpleResponse {
+	return (*Index)(ri).ReplicateDeletion(ctx, shardName, requestID, uuid)
+}
+
+func (ri *replicatedIndex) ReplicateDeletions(ctx context.Context, shardName,
```
[the remark this answers is not in the document]
Comment (parkerduckworth) [reply adapters/repos/db/index.go:1335]: Agreed, we don't need this anymore

File: usecases/sharding/replicated_index.go (hunk 1 above)
Comment (parkerduckworth) [reply usecases/sharding/replicated_index.go:48]: Ok, moved

File: entities/replica/error.go [hunk 4]
```diff
@@ -26,3 +26,8 @@ func (e *Error) Unwrap() error {
 func (e *Error) Error() string {
 	return fmt.Sprintf("%v :%v", e.Msg, e.Err)
```
[the remark this answers is not in the document]
Comment (parkerduckworth) [reply entities/replica/error.go:27]: Agreed, I have moved the `sharding/replicated_index` to `repos/db`, and `entities/replica` has been removed, with its contents restored to `usecases/replica`

File: adapters/repos/db/replication.go [hunk 5]
```diff
@@ -20,95 +20,148 @@ import (
...
+}
 
-	// TODO: metrics
-	s, err := NewShard(ctx, nil, shardName, ind)
-	if err != nil {
-		return err
+func (rii *ReplicatedIndex) ReplicateObject(ctx context.Context, indexName,
+	shardName, requestID string, object *storobj.Object,
+) replica.SimpleResponse {
+	index := rii.repo.GetReplicatedIndex(schema.ClassName(indexName))
```
Comment (redouan-rhazouani) [inline adapters/repos/db/replication.go:66]: Probably we can even get rid of ReplicatedIndex and GetReplicatedIndex in favour of DB.
I suggest instead to let DB implement the replication interface (see example below).

```
func (db *DB) ReplicateUpdate(ctx context.Context, indexName,
	shardName, requestID string, mergeDoc *objects.MergeDocument,
) replica.SimpleResponse {
	index := db.GetIndex(schema.ClassName(indexName))
	if index == nil {
		return replica.SimpleResponse{
			Errors: []string{fmt.Sprintf("local index %q not found", indexName)},
		}
	}

	return index.ReplicateUpdate(ctx, shardName, requestID, mergeDoc)
}


```

**Review by redouan-rhazouani — APPROVED** [review]

**Pull request merged**